File: Sources/VIPSImage+Caching.m
//
//  VIPSImage+Caching.m
//  VIPSKit
//
//  Caching methods with explicit format control
//

#import "VIPSImage+Private.h"

@implementation VIPSImage (Caching)

- (NSData *)cacheDataWithError:(NSError **)error {
    // Default to lossless WebP
    return [self cacheDataWithFormat:VIPSImageFormatWebP quality:0 lossless:YES error:error];
}

- (NSData *)cacheDataWithFormat:(VIPSImageFormat)format
                        quality:(NSInteger)quality
                       lossless:(BOOL)lossless
                          error:(NSError **)error {
    void *buffer = NULL;
    size_t length = 0;
    int result = -1;

    switch (format) {
        case VIPSImageFormatJPEG:
            // JPEG doesn't support lossless
            result = vips_jpegsave_buffer(self.image, &buffer, &length, "Q", (int)quality, NULL);
            break;
        case VIPSImageFormatPNG:
            // PNG is always lossless
            result = vips_pngsave_buffer(self.image, &buffer, &length, NULL);
            break;
        case VIPSImageFormatWebP:
            if (lossless) {
                result = vips_webpsave_buffer(self.image, &buffer, &length, "lossless", TRUE, NULL);
            } else {
                result = vips_webpsave_buffer(self.image, &buffer, &length, "Q", (int)quality, NULL);
            }
            break;
        case VIPSImageFormatHEIF:
            result = vips_heifsave_buffer(self.image, &buffer, &length, "Q", (int)quality, NULL);
            break;
        case VIPSImageFormatAVIF:
            result = vips_heifsave_buffer(self.image, &buffer, &length,
                                          "Q", (int)quality,
                                          "compression", VIPS_FOREIGN_HEIF_COMPRESSION_AV1,
                                          NULL);
            break;
        case VIPSImageFormatJXL:
            if (lossless) {
                result = vips_jxlsave_buffer(self.image, &buffer, &length, "lossless", TRUE, NULL);
            } else {
                result = vips_jxlsave_buffer(self.image, &buffer, &length, "Q", (int)quality, NULL);
            }
            break;
        case VIPSImageFormatGIF:
            result = vips_gifsave_buffer(self.image, &buffer, &length, NULL);
            break;
        case VIPSImageFormatUnknown:
            if (error) {
                *error = [NSError errorWithDomain:VIPSErrorDomain
                                             code:-1
                                         userInfo:@{NSLocalizedDescriptionKey: @"Unknown format for cache export"}];
            }
            return nil;
    }

    if (result != 0 || !buffer) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    NSData *data = [NSData dataWithBytesNoCopy:buffer length:length freeWhenDone:YES];
    return data;
}

- (BOOL)writeToCacheFile:(NSString *)path error:(NSError **)error {
    // Default to lossless WebP
    return [self writeToCacheFile:path format:VIPSImageFormatWebP quality:0 lossless:YES error:error];
}

- (BOOL)writeToCacheFile:(NSString *)path
                  format:(VIPSImageFormat)format
                 quality:(NSInteger)quality
                lossless:(BOOL)lossless
                   error:(NSError **)error {

    // Ensure correct extension for the format
    NSString *finalPath = path;
    NSString *expectedExt = [self extensionForFormat:format];
    if (expectedExt && ![[path.pathExtension lowercaseString] isEqualToString:expectedExt]) {
        finalPath = [path stringByAppendingPathExtension:expectedExt];
    }

    int result = -1;

    switch (format) {
        case VIPSImageFormatJPEG:
            result = vips_jpegsave(self.image, finalPath.UTF8String, "Q", (int)quality, NULL);
            break;
        case VIPSImageFormatPNG:
            result = vips_pngsave(self.image, finalPath.UTF8String, NULL);
            break;
        case VIPSImageFormatWebP:
            if (lossless) {
                result = vips_webpsave(self.image, finalPath.UTF8String, "lossless", TRUE, NULL);
            } else {
                result = vips_webpsave(self.image, finalPath.UTF8String, "Q", (int)quality, NULL);
            }
            break;
        case VIPSImageFormatHEIF:
            result = vips_heifsave(self.image, finalPath.UTF8String, "Q", (int)quality, NULL);
            break;
        case VIPSImageFormatAVIF:
            result = vips_heifsave(self.image, finalPath.UTF8String,
                                   "Q", (int)quality,
                                   "compression", VIPS_FOREIGN_HEIF_COMPRESSION_AV1,
                                   NULL);
            break;
        case VIPSImageFormatJXL:
            if (lossless) {
                result = vips_jxlsave(self.image, finalPath.UTF8String, "lossless", TRUE, NULL);
            } else {
                result = vips_jxlsave(self.image, finalPath.UTF8String, "Q", (int)quality, NULL);
            }
            break;
        case VIPSImageFormatGIF:
            result = vips_gifsave(self.image, finalPath.UTF8String, NULL);
            break;
        case VIPSImageFormatUnknown:
            if (error) {
                *error = [NSError errorWithDomain:VIPSErrorDomain
                                             code:-1
                                         userInfo:@{NSLocalizedDescriptionKey: @"Unknown format for cache export"}];
            }
            return NO;
    }

    if (result != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return NO;
    }

    return YES;
}

#pragma mark - Private

- (NSString *)extensionForFormat:(VIPSImageFormat)format {
    switch (format) {
        case VIPSImageFormatJPEG: return @"jpg";
        case VIPSImageFormatPNG: return @"png";
        case VIPSImageFormatWebP: return @"webp";
        case VIPSImageFormatHEIF: return @"heic";
        case VIPSImageFormatAVIF: return @"avif";
        case VIPSImageFormatJXL: return @"jxl";
        case VIPSImageFormatGIF: return @"gif";
        case VIPSImageFormatUnknown: return nil;
    }
    return nil;
}

@end

File: Sources/VIPSImage+Caching.swift
import Foundation
@_implementationOnly import vips
@_implementationOnly import CVIPS

extension VIPSImage {

    /// Export as lossless WebP for disk caching (~30% smaller than PNG).
    public func cacheData() throws -> Data {
        return try cacheData(format: .webP, quality: 0, lossless: true)
    }

    /// Export for caching with explicit format control.
    public func cacheData(format: ImageFormat, quality: Int = 0, lossless: Bool = true) throws -> Data {
        var buffer: UnsafeMutableRawPointer?
        var length: Int = 0
        let result: Int32

        switch format {
        case .jpeg:
            result = cvips_jpegsave_buffer(pointer, &buffer, &length, Int32(quality))
        case .png:
            result = cvips_pngsave_buffer(pointer, &buffer, &length)
        case .webP:
            if lossless {
                result = cvips_webpsave_buffer_lossless(pointer, &buffer, &length)
            } else {
                result = cvips_webpsave_buffer(pointer, &buffer, &length, Int32(quality))
            }
        case .heif:
            result = cvips_heifsave_buffer(pointer, &buffer, &length, Int32(quality))
        case .avif:
            result = cvips_avifsave_buffer(pointer, &buffer, &length, Int32(quality))
        case .jxl:
            if lossless {
                result = cvips_jxlsave_buffer_lossless(pointer, &buffer, &length)
            } else {
                result = cvips_jxlsave_buffer(pointer, &buffer, &length, Int32(quality))
            }
        case .gif:
            result = cvips_gifsave_buffer(pointer, &buffer, &length)
        case .unknown:
            throw VIPSError("Unknown format for cache export")
        }

        guard result == 0, let buffer else { throw VIPSError.fromVips() }
        defer { g_free(buffer) }
        return Data(bytes: buffer, count: length)
    }

    /// Write as lossless WebP cache file (auto-appends .webp extension).
    public func writeToCache(file path: String) throws {
        try writeToCache(file: path, format: .webP, quality: 0, lossless: true)
    }

    /// Write cache file with explicit format control.
    public func writeToCache(file path: String, format: ImageFormat, quality: Int = 0, lossless: Bool = true) throws {
        var finalPath = path
        if let ext = format.fileExtension,
           (path as NSString).pathExtension.lowercased() != ext {
            finalPath = (path as NSString).appendingPathExtension(ext) ?? path
        }

        let result: Int32
        switch format {
        case .jpeg:    result = cvips_jpegsave(pointer, finalPath, Int32(quality))
        case .png:     result = cvips_pngsave(pointer, finalPath)
        case .webP:
            if lossless {
                result = cvips_webpsave_lossless(pointer, finalPath)
            } else {
                result = cvips_webpsave(pointer, finalPath, Int32(quality))
            }
        case .heif:    result = cvips_heifsave(pointer, finalPath, Int32(quality))
        case .avif:    result = cvips_avifsave(pointer, finalPath, Int32(quality))
        case .jxl:
            if lossless {
                result = cvips_jxlsave_lossless(pointer, finalPath)
            } else {
                result = cvips_jxlsave(pointer, finalPath, Int32(quality))
            }
        case .gif:     result = cvips_gifsave(pointer, finalPath)
        case .unknown: throw VIPSError("Unknown format for cache export")
        }

        guard result == 0 else { throw VIPSError.fromVips() }
    }
}
